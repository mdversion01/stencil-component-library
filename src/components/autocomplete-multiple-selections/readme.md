# autocomplete-multiple-selections



<!-- Auto Generated Below -->


## Properties

| Property                  | Attribute                     | Description                                                                      | Type                             | Default                      |
| ------------------------- | ----------------------------- | -------------------------------------------------------------------------------- | -------------------------------- | ---------------------------- |
| `addBtn`                  | `add-btn`                     |                                                                                  | `boolean`                        | `false`                      |
| `addIcon`                 | `add-icon`                    |                                                                                  | `string`                         | `''`                         |
| `addNewOnEnter`           | `add-new-on-enter`            |                                                                                  | `boolean`                        | `true`                       |
| `arialabelledBy`          | `arialabelled-by`             |                                                                                  | `string`                         | `''`                         |
| `autoSort`                | `auto-sort`                   |                                                                                  | `boolean`                        | `true`                       |
| `badgeInlineStyles`       | `badge-inline-styles`         |                                                                                  | `string`                         | `''`                         |
| `badgeShape`              | `badge-shape`                 |                                                                                  | `string`                         | `''`                         |
| `badgeVariant`            | `badge-variant`               |                                                                                  | `string`                         | `''`                         |
| `clearIcon`               | `clear-icon`                  |                                                                                  | `string`                         | `''`                         |
| `clearInputOnBlurOutside` | `clear-input-on-blur-outside` |                                                                                  | `boolean`                        | `false`                      |
| `devMode`                 | `dev-mode`                    |                                                                                  | `boolean`                        | `false`                      |
| `disabled`                | `disabled`                    |                                                                                  | `boolean`                        | `false`                      |
| `editable`                | `editable`                    |                                                                                  | `boolean`                        | `false`                      |
| `error`                   | `error`                       |                                                                                  | `boolean`                        | `false`                      |
| `errorMessage`            | `error-message`               |                                                                                  | `string`                         | `''`                         |
| `formId`                  | `form-id`                     |                                                                                  | `string`                         | `''`                         |
| `formLayout`              | `form-layout`                 |                                                                                  | `"" \| "horizontal" \| "inline"` | `''`                         |
| `inputCol`                | `input-col`                   |                                                                                  | `number`                         | `10`                         |
| `inputCols`               | `input-cols`                  |                                                                                  | `string`                         | `''`                         |
| `inputId`                 | `input-id`                    |                                                                                  | `string`                         | `''`                         |
| `label`                   | `label`                       |                                                                                  | `string`                         | `''`                         |
| `labelCol`                | `label-col`                   |                                                                                  | `number`                         | `2`                          |
| `labelCols`               | `label-cols`                  | Responsive columns (e.g., "col", "col-sm-3 col-md-4", "xs-12 sm-6 md-4")         | `string`                         | `''`                         |
| `labelHidden`             | `label-hidden`                |                                                                                  | `boolean`                        | `false`                      |
| `labelSize`               | `label-size`                  |                                                                                  | `"" \| "lg" \| "sm"`             | `''`                         |
| `name`                    | `name`                        | Field name for selected items; if it ends with [] one input per item is emitted. | `string`                         | `undefined`                  |
| `options`                 | `options`                     |                                                                                  | `string[]`                       | `[]`                         |
| `placeholder`             | `placeholder`                 |                                                                                  | `string`                         | `'Type to search/filter...'` |
| `preserveInputOnSelect`   | `preserve-input-on-select`    |                                                                                  | `boolean`                        | `undefined`                  |
| `rawInputName`            | `raw-input-name`              | Also submit whatever the user typed under this name (verbatim).                  | `string`                         | `undefined`                  |
| `removeBtnBorder`         | `remove-btn-border`           |                                                                                  | `boolean`                        | `false`                      |
| `removeClearBtn`          | `remove-clear-btn`            |                                                                                  | `boolean`                        | `false`                      |
| `required`                | `required`                    |                                                                                  | `boolean`                        | `false`                      |
| `size`                    | `size`                        |                                                                                  | `"" \| "lg" \| "sm"`             | `''`                         |
| `type`                    | `type`                        |                                                                                  | `string`                         | `''`                         |
| `validation`              | `validation`                  |                                                                                  | `boolean`                        | `false`                      |
| `validationMessage`       | `validation-message`          |                                                                                  | `string`                         | `''`                         |


## Events

| Event               | Description | Type                                                |
| ------------------- | ----------- | --------------------------------------------------- |
| `clear`             |             | `CustomEvent<void>`                                 |
| `componentError`    |             | `CustomEvent<{ message: string; stack?: string; }>` |
| `itemSelect`        |             | `CustomEvent<string>`                               |
| `multiSelectChange` |             | `CustomEvent<string[]>`                             |
| `optionDelete`      |             | `CustomEvent<string>`                               |


## Methods

### `navigateOptions(direction: number) => Promise<void>`



#### Parameters

| Name        | Type     | Description |
| ----------- | -------- | ----------- |
| `direction` | `number` |             |

#### Returns

Type: `Promise<void>`




----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
