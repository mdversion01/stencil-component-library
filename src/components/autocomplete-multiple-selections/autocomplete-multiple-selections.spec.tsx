import { newSpecPage } from '@stencil/core/testing';
import { AutocompleteMultipleSelections } from './autocomplete-multiple-selections';

describe('autocomplete-multiple-selections', () => {
  it('renders with default props', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultipleSelections],
      html: `<autocomplete-multiple-selections label="Fruits" input-id="fruit"></autocomplete-multiple-selections>`,
    });
    expect(page.root).toBeTruthy();
    expect(page.root.querySelector('input')).toBeTruthy();
    expect(page.root.querySelector('label')?.textContent).toContain('Fruits');
    expect(page.root).toMatchSnapshot();
  });

  it('filters and renders dropdown with options', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultipleSelections],
      html: `<autocomplete-multiple-selections label="Fruit" input-id="fruit-id"></autocomplete-multiple-selections>`,
    });
    const comp = page.rootInstance;
    comp.options = ['Apple', 'Banana', 'Orange'];
    comp.inputValue = 'a';
    comp['filterOptions']();
    await page.waitForChanges();
    expect(comp.filteredOptions).toEqual(['Apple', 'Banana', 'Orange']);
    expect(page.root).toMatchSnapshot();
  });

  it('selects multiple items and renders selected badges', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultipleSelections],
      html: `<autocomplete-multiple-selections></autocomplete-multiple-selections>`,
    });
    const comp = page.rootInstance;
    comp.options = ['One', 'Two', 'Three'];
    comp.toggleItem('One');
    comp.toggleItem('Two');
    await page.waitForChanges();
    expect(comp.selectedItems).toEqual(['One', 'Two']);
    expect(page.root).toMatchSnapshot();
  });

  it('validates as invalid when required and nothing selected', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultipleSelections],
      html: `<autocomplete-multiple-selections required validation-message="Select at least one"></autocomplete-multiple-selections>`,
    });
    const comp = page.rootInstance;
    const valid = comp.validate();
    await page.waitForChanges();
    expect(valid).toBe(false);
    expect(comp.validation).toBe(true);
    expect(page.root).toMatchSnapshot();
  });

  it('handles keyboard navigation and enter selection', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultipleSelections],
      html: `<autocomplete-multiple-selections></autocomplete-multiple-selections>`,
    });
    const comp = page.rootInstance as AutocompleteMultipleSelections;
    comp.options = ['Alpha', 'Beta', 'Gamma'];
    comp.inputValue = 'a';
    comp.filterOptions();
    await page.waitForChanges();

    const input = page.root.querySelector('input');
    input.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowDown' }));
    await page.waitForChanges();
    input.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter' }));
    await page.waitForChanges();

    expect(comp.selectedItems).toContain('Alpha');
  });

  it('assigns correct aria attributes and roles', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultipleSelections],
      html: `<autocomplete-multiple-selections input-id="fruit" label="Fruits" arialabelled-by="label-id" required></autocomplete-multiple-selections>`,
    });

    const input = page.root.querySelector('input');
    expect(input.getAttribute('role')).toBe('combobox');
    expect(input.getAttribute('aria-autocomplete')).toBe('list');
    expect(input.getAttribute('aria-expanded')).toBe('false');
    expect(input.getAttribute('aria-controls')).toBe('fruit-listbox');
    expect(input.getAttribute('aria-labelledby')).toBe('label-id');
    expect(input.getAttribute('aria-required')).toBe('true');
    expect(input.getAttribute('aria-haspopup')).toBe('listbox');
  });

  it('handles Escape key to close dropdown', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultipleSelections],
      html: `<autocomplete-multiple-selections></autocomplete-multiple-selections>`,
    });

    const comp = page.rootInstance;
    comp.options = ['One', 'Two'];
    comp.inputValue = 'o';
    comp.filterOptions();
    await page.waitForChanges();

    const input = page.root.querySelector('input');
    input.dispatchEvent(new KeyboardEvent('keydown', { key: 'Escape' }));
    await page.waitForChanges();

    expect(comp.filteredOptions.length).toBe(0);
  });

  it('handles Backspace with no selection', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultipleSelections],
      html: `<autocomplete-multiple-selections></autocomplete-multiple-selections>`,
    });

    const comp = page.rootInstance;
    comp.inputValue = 'a';
    const input = page.root.querySelector('input');
    input.dispatchEvent(new KeyboardEvent('keydown', { key: 'Backspace' }));
    await page.waitForChanges();

    expect(comp.inputValue).toBe('a'); // Does not clear selection on Backspace
  });

  it('handles Tab to blur', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultipleSelections],
      html: `<autocomplete-multiple-selections></autocomplete-multiple-selections>`,
    });

    const comp = page.rootInstance;
    comp.isFocused = true;
    const input = page.root.querySelector('input');
    input.dispatchEvent(new KeyboardEvent('keydown', { key: 'Tab' }));
    comp.handleBlur();
    await page.waitForChanges();

    expect(comp.isFocused).toBe(false);
  });

  it('programmatically navigates and updates aria-activedescendant', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultipleSelections],
      html: `<autocomplete-multiple-selections input-id="id"></autocomplete-multiple-selections>`,
    });

    // ← cast to your component class, not the HTML interface
    const comp = page.rootInstance as AutocompleteMultipleSelections;

    comp.options = ['Apple', 'Banana', 'Cherry'];
    comp.inputValue = 'a';
    comp.filterOptions();
    await page.waitForChanges();

    const input: HTMLInputElement = page.root.shadowRoot ? page.root.shadowRoot.querySelector('input') : page.root.querySelector('input');

    comp.navigateOptions(1);
    await page.waitForChanges();

    expect(comp.focusedOptionIndex).toBe(0);
    expect(input.getAttribute('aria-activedescendant')).toBe('id-option-0');
  });

  it('actually wires onKeyDown to handleInput', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultipleSelections],
      html: `<autocomplete-multiple-selections input-id="id"></autocomplete-multiple-selections>`,
    });
    const comp = page.rootInstance as AutocompleteMultipleSelections;
    // spy
    const spy = jest.spyOn(comp as any, 'handleInput');

    // set up some options so the input even exists
    comp.options = ['A', 'B', 'C'];
    await page.waitForChanges();

    const input = page.root.shadowRoot ? page.root.shadowRoot.querySelector('input') : page.root.querySelector('input');

    input.dispatchEvent(
      new KeyboardEvent('keydown', {
        key: 'ArrowDown',
        bubbles: true,
        composed: true,
      }),
    );
    await page.waitForChanges();

    expect(spy).toHaveBeenCalled();
  });

  it('ignores Enter key when input does not match any option', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultipleSelections],
      html: `<autocomplete-multiple-selections></autocomplete-multiple-selections>`,
    });

    const comp = page.rootInstance;
    comp.options = ['One', 'Two', 'Three'];
    comp.inputValue = 'zzz';
    comp.filterOptions();
    await page.waitForChanges();

    const input = page.root.querySelector('input');
    input.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter' }));
    await page.waitForChanges();

    expect(comp.selectedItems).toEqual([]); // no match => nothing added
  });

  it('emits itemSelect when addBtn is clicked with inputValue', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultipleSelections],
      html: `<autocomplete-multiple-selections add-btn></autocomplete-multiple-selections>`,
    });

    const comp = page.rootInstance;
    const spy = jest.fn();
    page.root.addEventListener('itemSelect', spy);

    comp.inputValue = 'CustomOption';
    await page.waitForChanges();

    const btn = page.root.querySelector('button.add-btn') as HTMLButtonElement;
    btn.click();
    await page.waitForChanges();

    expect(spy).toHaveBeenCalledWith(expect.objectContaining({ detail: 'CustomOption' }));
  });

  it('disables input and buttons when disabled is true', async () => {
    const page = await newSpecPage({
      components: [AutocompleteMultipleSelections],
      html: `<autocomplete-multiple-selections disabled add-btn></autocomplete-multiple-selections>`,
    });

    const input = page.root.querySelector('input');
    const addBtn = page.root.querySelector('button.add-btn');
    expect(input.hasAttribute('disabled')).toBe(true);
    expect(addBtn.hasAttribute('disabled')).toBe(true);
  });
});
