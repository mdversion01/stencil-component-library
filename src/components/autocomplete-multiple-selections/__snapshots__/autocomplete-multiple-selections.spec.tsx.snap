// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`autocomplete-multiple-selections filters and renders dropdown with options 1`] = `
<autocomplete-multiple-selections input-id="fruit-id" label="Fruit">
  <div class="form-group">
    <div>
      <label class="form-label" htmlfor="fruit-Id">
        <span>
          Fruit
        </span>
      </label>
      <div>
        <div class="ac-multi-select-container">
          <div class="ac-selected-items"></div>
          <div class="ac-input-container">
            <div class="input-group">
              <input aria-autocomplete="list" aria-controls="fruit-Id-listbox" aria-expanded="true" aria-haspopup="listbox" aria-labelledby="" aria-required="false" autocomplete="off" class="form-control" id="fruit-Id" inputmode="text" name="fruit" placeholder="Type to search/filter..." role="combobox" spellcheck="false" type="" value="a">
              <button aria-label="Clear input" class="clear clear-btn input-group-btn" title="Clear input">
                <svg class="fa-icon" viewBox="0 0 384 512" xmlns="http://www.w3.org/2000/svg">
                  <path d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3l105.4 105.3c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256l105.3-105.4z"></path>
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div aria-live="polite" class="autocomplete-dropdown">
          <ul id="fruit-Id-listbox" role="listbox" tabindex="-1">
            <li aria-selected="false" class="autocomplete-dropdown-item" id="fruit-Id-option-0" role="option" tabindex="-1">
              <span>
                Apple
              </span>
            </li>
            <li aria-selected="false" class="autocomplete-dropdown-item" id="fruit-Id-option-1" role="option" tabindex="-1">
              <span>
                Banana
              </span>
            </li>
            <li aria-selected="false" class="autocomplete-dropdown-item" id="fruit-Id-option-2" role="option" tabindex="-1">
              <span>
                Orange
              </span>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</autocomplete-multiple-selections>
`;

exports[`autocomplete-multiple-selections renders with default props 1`] = `
<autocomplete-multiple-selections input-id="fruit" label="Fruits">
  <div class="form-group">
    <div>
      <label class="form-label" htmlfor="fruit">
        <span>
          Fruits
        </span>
      </label>
      <div>
        <div class="ac-multi-select-container">
          <div class="ac-selected-items"></div>
          <div class="ac-input-container">
            <div class="input-group">
              <input aria-autocomplete="list" aria-controls="fruit-listbox" aria-expanded="false" aria-haspopup="listbox" aria-labelledby="" aria-required="false" autocomplete="off" class="form-control" id="fruit" inputmode="text" name="fruits" placeholder="Type to search/filter..." role="combobox" spellcheck="false" type="" value="">
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</autocomplete-multiple-selections>
`;

exports[`autocomplete-multiple-selections selects multiple items and renders selected badges 1`] = `
<autocomplete-multiple-selections>
  <div class="form-group">
    <div>
      <div>
        <div class="ac-multi-select-container">
          <div class="ac-selected-items">
            <span class="badge text-bg-">
              <span>
                One
              </span>
              <button aria-label="Remove One" class="remove-btn" data-tag="One" role="button" title="Remove Tag">
                <svg viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg">
                  <path d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM175 175c9.4-9.4 24.6-9.4 33.9 0l47 47 47-47c9.4-9.4 24.6-9.4 33.9 0s9.4 24.6 0 33.9l-47 47 47 47c9.4 9.4 9.4 24.6 0 33.9s-24.6 9.4-33.9 0l-47-47-47 47c-9.4 9.4-24.6 9.4-33.9 0s-9.4-24.6 0-33.9l47-47-47-47c-9.4-9.4-9.4-24.6 0-33.9z"></path>
                </svg>
              </button>
            </span>
            <span class="badge text-bg-">
              <span>
                Two
              </span>
              <button aria-label="Remove Two" class="remove-btn" data-tag="Two" role="button" title="Remove Tag">
                <svg viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg">
                  <path d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM175 175c9.4-9.4 24.6-9.4 33.9 0l47 47 47-47c9.4-9.4 24.6-9.4 33.9 0s9.4 24.6 0 33.9l-47 47 47 47c9.4 9.4 9.4 24.6 0 33.9s-24.6 9.4-33.9 0l-47-47-47 47c-9.4 9.4-24.6 9.4-33.9 0s-9.4-24.6 0-33.9l47-47-47-47c-9.4-9.4-9.4-24.6 0-33.9z"></path>
                </svg>
              </button>
            </span>
          </div>
          <div class="ac-input-container">
            <div class="input-group">
              <input aria-autocomplete="list" aria-controls="-listbox" aria-expanded="true" aria-haspopup="listbox" aria-labelledby="" aria-required="false" autocomplete="off" class="form-control" inputmode="text" placeholder="Type to search/filter..." role="combobox" spellcheck="false" type="" value="">
              <button aria-label="Clear input" class="clear clear-btn input-group-btn" title="Clear input">
                <svg class="fa-icon" viewBox="0 0 384 512" xmlns="http://www.w3.org/2000/svg">
                  <path d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3l105.4 105.3c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256l105.3-105.4z"></path>
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</autocomplete-multiple-selections>
`;

exports[`autocomplete-multiple-selections validates as invalid when required and nothing selected 1`] = `
<autocomplete-multiple-selections required="" validation-message="Select at least one">
  <div class="form-group">
    <div>
      <div>
        <div class="ac-multi-select-container is-invalid">
          <div class="ac-selected-items"></div>
          <div class="ac-input-container">
            <div class="input-group">
              <input aria-autocomplete="list" aria-controls="-listbox" aria-describedby="-validation" aria-expanded="false" aria-haspopup="listbox" aria-labelledby="" aria-required="true" autocomplete="off" class="form-control is-invalid" inputmode="text" placeholder="Type to search/filter..." role="combobox" spellcheck="false" type="" value="">
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="invalid-feedback">
      Select at least one
    </div>
  </div>
</autocomplete-multiple-selections>
`;
